ibridges/irodsconnector/tickets.py

```diff
@@ -43,13 +46,72 @@ def create_ticket(self, obj_path: str, expiry_string: str = '') -> tuple:
             (str, bool)
 
         """
-        ticket = irods.ticket.Ticket(self.sess_man.irods_session)
-        ticket.issue('read', obj_path)
-        logging.info('CREATE TICKET: %s: %s', ticket.ticket, obj_path)
+        ticket = irods.ticket.Ticket(self.session.irods_session)
+        ticket.issue(ticket_type, obj_path)
         expiration_set = False
-        if expiry_string != '':
+        if expiry_string is not None:
             try:
                 expiration_set = ticket.modify('expire', expiry_string) == ticket
             except Exception as error:
-                logging.error('Could not set expiration date: %r', error)
+                self.delete_ticket(ticket)
+                raise Exception('Could not set expiration date: %r', error)
         return ticket.ticket, expiration_set
+
+    @property
+    def all_ticket_strings(self) -> list:
+        return [name for name, _, _, _ in self.all_tickets()]
+
+    def get_ticket(self, ticket_str: str) -> irods.ticket.Ticket:
+        if ticket_str in self.all_ticket_strings:
+            return irods.ticket.Ticket(self.session.irods_session, ticket=ticket_str)
+        else:
+            warnings.warn("Ticket does not exist.")
+        return None
+
+    def delete_ticket(self, ticket: irods.ticket.Ticket):
+        if ticket and ticket.string in self.all_ticket_strings:
+            ticket.delete()
+            self.all_tickets(update=True)
+        else:
+            warnings.warn("Ticket does not exist.")
+
+    def all_tickets(self, update: bool = False) -> list:
+        """retrieves all tickets and their metadata belonging to the user.
+
+        Parameters
+        ----------
+        update : bool
+            Refresh information from server.
+
+        Returns
+        -------
+        list
+            [(ticket string, ticket type, irods obj/coll id, expiry data in epoche)]
+        """
+        user = self.session.username
+        if update or self._all_tickets is None:
+            self._all_tickets = []
+            for row in self.session.irods_session.query(TicketQuery.Ticket).filter(
+                    TicketQuery.Owner.name == user):
+                self._all_tickets.append((row[TicketQuery.Ticket.string],
+                                          row[TicketQuery.Ticket.type],
+                                          self._id_to_path(str(row[TicketQuery.Ticket.object_id])),
+                                          row[TicketQuery.Ticket.expiry_ts]
+                                          ))
+        return self._all_tickets
+
+    def _id_to_path(self, itemid: str) -> str:
```

Comment:
Can you add some comments in this function? If I understand correctly, the first `if` is for data objects, and the second part for collections? The itemid is the id that is assigned to a ticket, or is it a general id? If it is specific to tickets, then maybe ticket_id might be more descriptive?

Reply:
I added some explanation. Will this work?

    """
    Given an iRODS item id (data object or collection) from the
    TicketQuery.Ticket.object_id this function retrieves the corresponding
    iRODS path.
    """

Reply:
Yes!